data-model/src/groupings/range.rs

```diff
@@ -0,0 +1,159 @@
+        Some(self.cmp(other))
+    }
+}
+
+/// One-dimensional grouping over a type of value.
+///
+/// [Definition](https://willowprotocol.org/specs/grouping-entries/index.html#range)
+#[derive(Debug, PartialEq, Eq)]
+pub struct Range<T>
+where
+    T: PartialEq + Eq + PartialOrd + Ord,
+{
+    /// A range [includes](https://willowprotocol.org/specs/grouping-entries/index.html#range_include) all values greater than or equal to its [start value](https://willowprotocol.org/specs/grouping-entries/index.html#start_value).
+    pub start: T,
+    /// A range [includes](https://willowprotocol.org/specs/grouping-entries/index.html#range_include) all values strictly less than its end value (if it is not open).
+    pub end: RangeEnd<T>,
+}
+
+/// An error indicating a range is [empty](https://willowprotocol.org/specs/grouping-entries/index.html#range_empty) and will never [include](https://willowprotocol.org/specs/grouping-entries/index.html#range_include) any values.
+#[derive(Debug)]
+pub struct EmptyRangeError;
+
+impl<T> Range<T>
+where
+    T: PartialEq + Eq + PartialOrd + Ord + Clone,
+{
+    /// Construct a new [open range](https://willowprotocol.org/specs/grouping-entries/index.html#open_range) from a [start value](https://willowprotocol.org/specs/grouping-entries/index.html#start_value).
+    pub fn new_open(start: T) -> Self {
+        Self {
+            start,
+            end: RangeEnd::Open,
+        }
+    }
+
+    /// Construct a new [closed range](https://willowprotocol.org/specs/grouping-entries/index.html#closed_range) from a [start](https://willowprotocol.org/specs/grouping-entries/index.html#start_value) and [end_value](https://willowprotocol.org/specs/grouping-entries/index.html#end_value), or a [`UselessRangeError`] if the resulting range would never [include](https://willowprotocol.org/specs/grouping-entries/index.html#range_include) any values.
+    pub fn new_closed(start: T, end: T) -> Result<Self, EmptyRangeError> {
+        if start < end {
+            return Ok(Self {
+                start,
+                end: RangeEnd::Closed(end),
+            });
+        }
+
+        Err(EmptyRangeError)
+    }
+
+    /// Return whether a given value is [included](https://willowprotocol.org/specs/grouping-entries/index.html#range_include) by the [`Range`] or not.
+    pub fn includes(&self, value: T) -> bool {
+        self.start <= value && self.end.gt_val(&value)
+    }
+
+    /// Returns `true` if `other` range is fully [included](https://willowprotocol.org/specs/grouping-entries/index.html#range_include) in this [`Range`].
+    pub fn includes_range(&self, other: &Range<T>) -> bool {
+        self.start <= other.start && self.end >= other.end
+    }
+
+    /// Create the [intersection](https://willowprotocol.org/specs/grouping-entries/index.html#range_intersection) between this [`Range`] and another `Range`.
+    pub fn intersection(&self, other: &Self) -> Option<Self> {
+        let start = (&self.start).max(&other.start);
```

Comment:
```suggestion
        let start = cmp::max(&self.start, &other.start);
```

Is more idiomatic I think.